src/ape/contracts/base.py

```diff
@@ -202,34 +201,196 @@ def __call__(self, *args, **kwargs) -> ReceiptAPI:
             abi=selected_abi,
             address=self.contract.address,
         )(*args, **kwargs)
 
 
-@dataclass
-class ContractLog:
-    name: str
-    data: Dict[str, Any]
+class ContractEvent(ManagerAccessMixin):
+    """
+    The types of events on a :class:`~ape.contracts.base.ContractInstance`.
+    Use the event types via ``.`` access on the contract instances.
 
+    Usage example::
+
+         # 'my_contract' refers to a ContractInstance in this case.
+         my_event_type = my_contract.MyEvent
+    """
 
-class ContractEvent(ManagerAccessMixin):
     def __init__(
         self,
         contract: "ContractInstance",
-        abis: List[EventABI],
+        abi: EventABI,
         cached_logs: List[ContractLog] = None,
     ) -> None:
         super().__init__()
         self.contract = contract
-        self.abis = abis
+        self.abi = abi
         self.cached_logs = cached_logs or []
 
+    @property
+    def name(self) -> str:
+        """
+        The name of the contract event, as defined in the contract.
+        """
+
+        return self.abi.name
+
+    def __iter__(self) -> Iterator[ContractLog]:
+        yield from self.search()
+
+    @singledispatchmethod
+    def __getitem__(self, value) -> Union[ContractLog, List[ContractLog]]:
+        raise NotImplementedError(f"Cannot use '{type(value)}' to access logs.")
+
+    @__getitem__.register
+    def __getitem_int(self, index: int) -> ContractLog:
+        """
+        Access events on the contract by the index of when they occurred.
+
+        Args:
+            index (int): The index such that ``0`` is the first log to have occurred
+              and ``-1`` is the last.
+
+        Returns:
+            :class:`~ape.contracts.base.ContractLog`
+        """
+
+        if index == 0:
+            logs_slice = [next(self._get_logs_iter())]
+        elif index > 0:
+            # Call over to 'self.__getitem_slice'.
+            logs_slice = self[: index + 1]  # type: ignore
+        else:
+            # Call over to 'self.__getitem_slice'.
+            logs_slice = self[index:]  # type: ignore
+
+        try:
+            return logs_slice[index]
+        except IndexError as err:
+            raise IndexError(f"No log at index '{index}' for event '{self.abi.name}'.") from err
+
+    @__getitem__.register
+    def __getitem_slice(self, value: slice) -> List[ContractLog]:
+        start = value.start or 0
+        stop = value.stop if value.stop is not None else self.chain_manager.blocks.height
+        step = value.step or 1
+        collected_logs: List[ContractLog] = []
+        counter = 0
+        for log in self._get_logs_iter():
+            if counter < start:
+                counter += 1
+                continue
+
+            elif counter >= stop:
+                return collected_logs
+
+            elif counter >= start:
+                collected_logs.append(log)
+                counter += step
+
+        return collected_logs
+
+    def search(
```

Comment:
I know it's doing a search here, but the name doesn't match other similar functions we have. I'd like to create a unified API for accessing query functionality across all objects like this so people don't have to remember special syntax across all the common UX points

Reply:
What are some other similar functions? Do you mean like `range()` f rom `BlockContainer`?

We can go back to `filter()`

Reply:
Or maybe... We can make this behave like `range()`